doc/probabel.1
.TH ProbABEL 1 "4 May 2016" "ProbABEL 0.5.0"
.SH NAME
probabel \- Wrapper around the three ProbABEL binaries, simplifying their use
.SH SYNOPSIS
.B probabel
.BI "" "CHROM-START CHROM-STOP METHOD COHORT MODEL PHENOTYPE [OTHER_OPTIONS]"
.SH DESCRIPTION
.B probabel
is a wrapper around the three ProbABEL binaries: \fBpalinear\fR,
\fBpalogist\fR, and \fBpacoxph\fR.
.PP
It is designed to simplify runnig of a Genome-Wide Association Study
(GWAS) by saving the user's precious time in two ways:
.PP
1) it runs the regression analysis of all chromosomes from \fICHROM-START\fR to
\fICHROM-STOP\fR.
.PP
2) In contrast to the three binaries mentioned above,
with \fBprobabel\fR the user doesn't need to specify the locations
of the required genetic data files. Their location is centrally
managed in a configuration file (\fI/etc/probabel_config.cfg\fR by default).

.SH OPTIONS
.SS Required command line options
.TP
.I CHROM-START
Number of the chromosome to start the analysis at.
.TP
.I CHROM-END
Number of the chromosome to end the analysis at.
.PP
Note that analysis of a single chromosome can be run by specifying the
same number for \fICHROM-START\fR and \fICHROM-STOP\fR. In fact, this
is the way to run analysis for the X or Y chromosome
.TP
.I METHOD
Which regression method to use, can be one of \fIlinear\fR,
\fIlogistic,\fR or \fIcoxph\fR.

.TP
.I COHORT
The name of the cohort as defined in the
\fI/etc/probabel_config.cfg\fR file. Using the cohort identifier from
the config file the three actual ProbABEL binaries know where to find
the genotypic information files (dosage files and/or probability
files).
.TP
.I MODEL
The genetic model to be used. Can be either \fB--additive\fR for the
additive model (requires genotype data in dosage form), or
\fB--allmodels\fR to run all genetic models: additive, recessive,
dominant, over-dominant and 2df.
.TP
.I PHENOTYPE
The name of the phenotype file, without its extension (which should be
\fI.PHE\fR!).

.SS Other options
Any options listed after the (mandatory) previous ones will be passed
on to the underlying binary: \fBpalinear\fR, \fBpalogist\fR or
\fBpacoxph\fR, depending on the \fIMETHOD\fR.
.PP
The most commonly used option is the \fB\-o\fR option, followed by a
file name, this specifies the beginning of the file name for the
output files.

.SS Additional options
.B "\-h, \-\-help"
Get additional help on how to run \fBprobabel\fR. In this case none of
the other options needs to be specified.

.SH EXAMPLES
This runs a genome-wide association scan using linear regression and
the additive genetic model on cohort \fIMy_Large_Cohort\fR for
phenotype \fImy_phenotype\fR:
.PP
.nf
.RS
probabel 1 22 linear "My_Large_Cohort" \-\-additive my_phenotype
.RE
.fi
.PP

.SH "SEE ALSO"
palinear(1), palogist(1), pacoxph(1)
.SH BUGS
The bugtracker is located at
.br
https://github.com/GenABEL-Project/ProbABEL/issues
.SH AUTHORS
Lennart C. Karssen


doc/ChangeLog
***** v.0.5.0 (2016.)
* ProbABEL now depends on the Eigen library. The option to compile without
  Eigen has been removed.
* Fixed Issue #15: Coxph interactions not working; when using the
  --interaction option the name of the model was wrong (e.g. for
  --interaction=1, the interaction term was shown as mu * SNP_A1, which
  obviously is not what we meant to do).
* Fixed issue Github #16: "Too many NANs for SNPs with low `Rsq` when some
  commandline options are specified"
* Added --flipmaf command line option to palinear, pacoxph and palogist.
  If this option is selected, ProbABEL will 'flip' the alleles A1 and A2
  for those variants for which the Freq1 value in the info file is less
  than 0.5. Changing the reference and effect allele in this way can
  sometimes (for some genetic models and low MAF), reduce regression
  errors and thus result in proper estimates which otherwise be set to
  NaN. Many thanks for the group of Prof. Kiemeney at the dept. of Health
  Evidence Radboud UMC in Nijmegen (NL) for sponsoring the work on this
  feature! Also thanks to Tessel Galesloot from the same group for
  extensive testing of this release.
* Don't set the betas and se_betas for all covariates in case Cox
  regression problems like singularities occur, but only the beta/se_beta
  of the corresponding covariate. Also make the corresponding warning
  messages more clear by including the name of the genetic model that is
  being tested.
* Not-a-number values in the output are now printed as "nan" instead of
  "NaN".
* Linear regression mmscore option is 2 times faster
* Speedup of Linear regression: (measured with multiple runs with 2 covariates,
  33815 SNP and 3485 people (using all) compared to v.0.4.3)
** Reading mldose/mlprob files 14 times faster
** Calculation of regression is more than 3.5 faster
** Overall runtime using above settings is 5 times faster
* Handles multiple variants of NaN (NA,Na,Nan,na,nan) correct while reading of
  mldose/mlprob files
* Fixed Issue #21: Remove "interaction_only" option. According to Yurii's
  post on the forum this option never worked as expected.
* Fixed bugs #5883 & #6010 (on R-forge): The main effect is displayed in
  the output with the  `interaction_only` option. Thanks to Maksim
  Struchalin & Lennart Karssen for fixing it and to Farid Radmanesh for
  reporting it. This was only a display error, computations were not
  affected.
* Fixed bug #5982: ProbABEL's make install fails on MacOS X and FreeBSD
  with sed error. Thanks to forum user mmold for reporting the bug.


***** v.0.4.5 (2015.05.26)
* Fixed bug #6054: "Not all ProbABEL's short command line options are
  correctly parsed."
* Fixed bug #6041: "ProbABEL's Cox module reports too many errors (beta
  may be infinite, setting beta and se to 'NaN')". Thanks to Anne
  Grotenhuis from the Radboud Medical Centre Nijmegen and forum user
  quentin300 for their extensive bug reports and help in testing the fix.
* Fixed bug #1266: "pacoxph with no covariates"; ProbABEL now also works
  when no covariates have been provided. Thanks to Aaron Joon for
  reporting this bug back in 2011 and to Anne Grotenhuis and to forum user
  quentin300 for pushing this bug to my attention and their help in beta
  testing.
* pacoxph now displays the regression equation correctly.
* Minor fix to the manual: the Rsq values in the info file should be >
  1e-16 (and not > 0 as mentioned before) otherwise the output will be set
  to 'nan'.
* Minor fix to the man-pages: according to the man-pages palogist,
  palinear and pacoxph all did regression using a linear model.


***** v.0.4.4 (2014.11.07)
* Fixed bug #5729 in the Cox PH module. Some checks for problems with the
  regression were incorrectly implemented. Thanks to Matthias Wuttke from
  the University Medical Centre Freiburg, Anne Grotenhuis from the Radboud
  Medical Centre Nijmegen, and Luba Pardo and Joris Verkouteren from the
  Erasmus Medical Centre Rotterdam for their time and effort in reporting
  the bug, helping to identify the problem and testing the fix.
* Backported fix of bug #5982: ProbABEL's make install fails on MacOS X
  and FreeBSD with sed error. Thanks to forum user mmold for reporting the bug.


***** v.0.4.3 (2014.04.01)
* Speed-up of a factor of ~ 2 for linear, logistic and Cox regression when
  using filevector input files.
* Fixed bug #5404: "ProbABEL's R check for Cox regression doesn't check if
  the survival package is installed".
* Fixed bug #5403: "The ProbABEL manual doesn't contain any information on
  how to install ProbABEL"


***** v.0.4.2 (2014.01.02)
* The 'probabel.pl' script is now simply renamed to 'probabel' (a user
  shouldn't care what scripting language we use). For at least several
  releases to come, the old script name will still exist (as a link to the
  original) and a warning message is displayed when the user runs the
  .pl script. This should give people time to adjust their pipelines.
* Fix bug #4919: Too small reading buffers for long alleles in mach info
  and legend files. Thanks to Daniel Taliun for reporting the bug and
  providing the patch. Thanks to Xia Shen for testing.
* Fix bug #4776: Specifying --sep="\t" as an option to pa* doesn't insert
  a TAB as separator. Now ProbABEL inserts proper tabs when specifying
  "\t". Thanks to Maksim Struchalin for fixing this bug.
* Improved convergence checks in the Cox PH regression module. The checks
  now give similar errors as R does.
* A minor change in the screen output of ProbABEL. Some of the status
  messages ("Reading phenotype data" etc.) have been added or move to a
  slightly different place in the code to help debugging problems with the
  input data.
* Fix a bug in the example scripts: an incorrect shell variable was used.
* The extract-snp binary is no longer built by default (as it isn't
  finished yet). Building can be enabled at compile time by giving the
  --enable-extract-snp option to ./configure.

src/probabel.pl
#! /usr/bin/perl -W
#==========================================================================
#
#           Filename:  probabel.pl
#
#        Description: Handy perl wrapper for ProbABEL functions
#
#==========================================================================
use strict;

#==========================================================================
# Set variables
my $version="PROBABEL_VERSION";

# Define some filename postfixes
my $_2df_file_postfix = "_2df.out.txt";
my $_add_file_postfix = "_add.out.txt";
my $_domin_file_postfix = "_domin.out.txt";
my $_recess_file_postfix = "_recess.out.txt";
my $_over_domin_file_postfix = "_over_domin.out.txt";

# Separators in the config file
my $separator_cfg = ",";
my $chr_replacement = "_._chr_._";
my $chunk_replacement = "_._chunk_._";

# Set file locations
my $base_path = "./";
my @anprog = ($base_path . "palinear",
              $base_path . "palogist",
              $base_path . "pacoxph");
my $config = "probabel_config.cfg";

# Define the regression methods that are implemented
my @method = ("linear", "logistic", "coxph");

my %cohorts;
my @mlinfos;
my @mldoses;
my @mlprobs;
my @legends;


#==========================================================================
# Read config file
open(CFG, "$config") or die "Reading configuration file $config failed: $!" .
    "\nDid you forget to edit and rename the probabel_config.cfg.example file?\n";

<CFG>; #skip the first line (header)

for(my $i=0 ; my $line = <CFG> ; $i++)
{
    chomp($line);
    next if ($line =~ /^#/);
    next if ($line =~ /^$/);
    my @line_array = split(/$separator_cfg/, $line);
    $cohorts{$line_array[0]} = $i;
    $mlinfos[$i]  = $line_array[1];
    $mldoses[$i]  = $line_array[2];
    $mlprobs[$i] = $line_array[3];
    $legends[$i]  = $line_array[4];
}
close(CFG);


#==========================================================================
# Print usage info if arguments are not correct
if(@ARGV<6 || $ARGV[0] eq "--help" || $ARGV[0] eq "-h") {
    print "Usage:
        probabel.pl chrom-start chrom-end method cohort <--allmodels OR --additive> trait <other available options of ProbABEL functions>\n";
    print "\n	* chrom-start - the first chromosome number, chrom-end - the last one; X or Y have to be run separately (specify them twice, once as chrom-start and once as chrom-end)";
    print "\n	* method can be ";
    foreach my $me(@method) {print "\"".$me."\", "};
    print "\n	* use --allmodels if you need dominant, recessive and heterozygous models
          and --additive if additive only\n";
    print "	* Available cohorts are ";
    foreach my $coh(keys %cohorts) {
        print "\"".$coh."\", "
    };
    print "\n	* example:
          probabel.pl 1 22 linear \"ERF\" --additive filename
          (filename has to be saved as filename.PHE)\n\n";

    if(@ARGV == 1 && ($ARGV[0] eq "--help" || $ARGV[0] eq "-h")) {
        print "\nDetails:\n";
        print " The probabel.pl script is used for analysis of imputed data. First you have to create a file with the phenotype values that you are going to use. The first column contains ids in special order, the second one contains the trait which you are going	analyze, the others contain covariates.  For example:

        id         phen1 covariate1  covariate2
        1_2094  0     334         0
        1_5060  1     56          1
        1_4077  1     346         6
        .
        .
        .

        This implies the model:
        phen1 ~ covariate1 + covariate2 + SNP


        Then save it to folder where you are doing the analysis. The name of the file must be name_of_file.PHE, where name_of_file is any name.

        Then run the following on the command line:
          probabel.pl 1 22 \"method\" \"cohort\" --model name_of_file
        Change \"method\", \"cohort\" and --model to appropriate values\n";
        print "\n	Version: $version";
        print "\n\n	Authors: Lennart Karssen   - lennart\@karssen.org,
                 Maarten Kooyman - m.kooijman\@erasmusmc.nl,
                 Maksim Struchalin - m.struchalin\@erasmusmc.nl,
                 Yurii Aulchenko   - yurii.aulchenko\@gmail.com.\n\n";
    }
    else {
        print "Type probabel.pl --help for more details.\n";
    }
    exit;
}


#==========================================================================
# Put the command line arguments into variables and verify them
my $startchr = $ARGV[0];
my $endchr = $ARGV[1];
my $method = $ARGV[2];
my $chohort = $ARGV[3];
my $model = $ARGV[4];

die "error: chrom-start is > 22" if($startchr > 22 && $startchr != "X") ;
die "error: chrom-end is > 22" if($endchr > 22 && $endchr != "X");
die "error: chrom-start > chrom-end" if($startchr > $endchr);


my $cohort_position = $cohorts{$chohort};


if(!defined($cohort_position)) {
    print "\nerror: Wrong cohort name, \"$chohort\" is not an available cohort.
Available cohorts are ";
    foreach my $coh(keys %cohorts) {
        print "\"".$coh."\", ";
    }
    print "\n\n";
    exit;
}

my $mlinfo = $mlinfos[$cohort_position];
my $mldose = $mldoses[$cohort_position];
my $mlprob = $mlprobs[$cohort_position];
my $legend = $legends[$cohort_position];


my $passed = 0;
my $prog;
for (my $i=0; $i<@method; $i++) {
    if ($ARGV[2] eq $method[$i]) {
        $passed = 1;
        $prog = $anprog[$i];
    }
}
die "error: Wrong method. method has to be one of: @method\n" if (!$passed);


my $phename = $ARGV[5];
# By default the output file prefix is the same as the name of the
# phenotype file (minus the .PHE extension and any paths)
use File::Basename;
my $outfile_prefix = basename($phename);
my $keys="";
for (my $i=6; $i<@ARGV; $i++) {
    if ($ARGV[$i] eq "-o")
    {
        # Apparently the user wants to change the output file name
        $outfile_prefix = $ARGV[$i+1];

        # Skip the next argument (supposedly the addition to the
        # output file name).
        $i++;
    }
    else
    {
        $keys = $keys.$ARGV[$i]." ";
    }
}
chop($keys);

my $model_option_num = 0;
my $mldose_prob;
if($model eq "--additive") {
    $mldose_prob = $mldose;
    $model_option_num = 1;
} elsif($model eq "--allmodels") {
    $mldose_prob = $mlprob;
    $model_option_num = 2;
} else {
    die "error: Wrong key for model. You can use \"--additive\" or \"--allmodels\" only\n";
}


#==========================================================================
# Start the analysis now that the input has been validated
print "Start...\n";

my $chr = $startchr;
my $hadhead=0;
my $head;
my $mlinfo_arg;
my $mldose_arg;
my $legend_arg;
my $outfile_arg;

# Separate command for the sex chromosomes.
if ($chr eq "X" || $chr eq "Y") {
    $mlinfo_arg = $mlinfo;
    $mlinfo_arg =~ s/$chr_replacement/$chr/g;

    $mldose_arg = $mldose_prob;
    $mldose_arg =~ s/$chr_replacement/$chr/g;

    $legend_arg = $legend;
    $legend_arg =~ s/$chr_replacement/$chr/g;

    if($hadhead==0) {
        $head="";
        $hadhead=1;
    } else {
        my $head="--no-head";
    }

    system "$prog -p $phename.PHE --ngpreds $model_option_num -i $mlinfo_arg -d $mldose_arg -m $legend_arg --chrom $chr -o $outfile_prefix $head $keys";

    exit;
}

# Clean up any existing output files
if($model_option_num==2)
{
    system "rm ${outfile_prefix}${_2df_file_postfix} 2>/dev/null";
    system "rm ${outfile_prefix}${_add_file_postfix} 2>/dev/null";
    system "rm ${outfile_prefix}${_domin_file_postfix} 2>/dev/null";
    system "rm ${outfile_prefix}${_recess_file_postfix} 2>/dev/null";
    system "rm ${outfile_prefix}${_over_domin_file_postfix} 2>/dev/null";
} else {
    system "rm ${outfile_prefix}${_add_file_postfix} 2>/dev/null";
}

# Commands for the autosomes
for($chr=$startchr; $chr<=$endchr; $chr++) {

    my $nrchunks = 0;
    # Find out the number of chunks for the current chromosome
    my $infofiles = $mlinfo;
    $infofiles =~ s/$chr_replacement/$chr/g;
    $infofiles =~ s/$chunk_replacement/*/g;
    $nrchunks = `ls $infofiles 2>/dev/null | wc -l`;
    if ($nrchunks==0) {
        # If no chunked info files exist the 'wc -l' command returns 0
        # so that actually means 1 chunk containing all data.
        $nrchunks = 1;
    }
    print "Nr. of chunks: $nrchunks";

    # Loop over all chunks
    for (my $chunk=1; $chunk <= $nrchunks; $chunk++)
    {
        if($hadhead==0) {
            $head="";
            $hadhead=1;
        } else {
            $head="--no-head";
        }
        $mlinfo_arg = $mlinfo;
        $mlinfo_arg =~ s/$chr_replacement/$chr/g;
        $mlinfo_arg =~ s/$chunk_replacement/$chunk/g;

        $mldose_arg = $mldose_prob;
        $mldose_arg =~ s/$chr_replacement/$chr/g;
        $mldose_arg =~ s/$chunk_replacement/$chunk/g;

        $legend_arg = $legend;
        $legend_arg =~ s/$chr_replacement/$chr/g;
        $legend_arg =~ s/$chunk_replacement/$chunk/g;

        $outfile_arg = "$outfile_prefix.chunk${chunk}.chr${chr}";

        my $command = "$prog -p $phename.PHE --ngpreds $model_option_num ";
        $command = $command . "-i $mlinfo_arg -d $mldose_arg -m $legend_arg";
        $command = $command . " --chrom $chr";
        $command = $command . " -o $outfile_arg ";
        $command = $command . "$head $keys";
        print "$command \n";
        system $command;

        # Combine the output data for all chunks of this chromosome
        if($model_option_num==2)
        {
            `cat $outfile_arg$_2df_file_postfix >> ${outfile_prefix}.${chr}${_2df_file_postfix}`;
            `rm $outfile_arg$_2df_file_postfix`;

            `cat $outfile_arg$_add_file_postfix >> ${outfile_prefix}.${chr}${_add_file_postfix}`;
            `rm $outfile_arg$_add_file_postfix`;

            `cat $outfile_arg$_domin_file_postfix >> ${outfile_prefix}.${chr}${_domin_file_postfix}`;
            `rm $outfile_arg$_domin_file_postfix`;

            `cat $outfile_arg$_recess_file_postfix >> ${outfile_prefix}.${chr}${_recess_file_postfix}`;
            `rm $outfile_arg$_recess_file_postfix`;

            `cat $outfile_arg$_over_domin_file_postfix >> ${outfile_prefix}.${chr}${_over_domin_file_postfix}`;
            `rm $outfile_arg$_over_domin_file_postfix`;
        } else {
            `cat $outfile_arg$_add_file_postfix >> $outfile_prefix.${chr}${_add_file_postfix}`;
            print "cat $outfile_arg$_add_file_postfix >> $outfile_prefix.chr${chr}${_add_file_postfix}\n";
            `rm $outfile_arg$_add_file_postfix`;
            print "rm $outfile_arg$_add_file_postfix\n";
        }
    }

    # Combine the output data for all chromosomes into one file
    if($model_option_num==2)
    {
        `cat $outfile_prefix.${chr}$_2df_file_postfix >> ${outfile_prefix}${_2df_file_postfix}`;
        `rm $outfile_prefix.${chr}$_2df_file_postfix`;

        `cat $outfile_prefix.${chr}$_add_file_postfix >> ${outfile_prefix}${_add_file_postfix}`;
        `rm $outfile_prefix.${chr}$_add_file_postfix`;

        `cat $outfile_prefix.${chr}$_domin_file_postfix >> ${outfile_prefix}${_domin_file_postfix}`;
        `rm $outfile_prefix.${chr}$_domin_file_postfix`;

        `cat $outfile_prefix.${chr}$_recess_file_postfix >> ${outfile_prefix}${_recess_file_postfix}`;
        `rm $outfile_prefix.${chr}$_recess_file_postfix`;

        `cat $outfile_prefix.${chr}$_over_domin_file_postfix >> ${outfile_prefix}${_over_domin_file_postfix}`;
        `rm $outfile_prefix.${chr}$_over_domin_file_postfix`;
    } else {
        `cat $outfile_prefix.${chr}$_add_file_postfix >> ${outfile_prefix}${_add_file_postfix}`;
        print "cat $outfile_prefix.${chr}$_add_file_postfix >> ${outfile_prefix}${_add_file_postfix}\n";
        `rm $outfile_prefix.${chr}$_add_file_postfix`;
        print "rm $outfile_prefix.${chr}$_add_file_postfix\n";
    }
}

print "Finished all chromosomes ...\n";
